app/modules/websocket.c

```diff
@@ -0,0 +1,332 @@
+
+  if (data->onConnection != LUA_NOREF) {
+    NODE_DBG("got an onConnection func! !\n");
+
+    lua_rawgeti(L, LUA_REGISTRYINDEX, data->onConnection); // load the callback function
+    lua_rawgeti(L, LUA_REGISTRYINDEX, data->self_ref);  // pass itself, #1 callback argument
+    lua_call(L, 1, 0);
+  }
+}
+
+static void websocketclient_onReceiveCallback(ws_info *ws, char *message, int opCode) {
+  NODE_DBG("websocketclient_onReceiveCallback\n");
+
+  lua_State *L = lua_getstate();
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    luaL_error(L, "Client websocket is nil.\n");
+    return;
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (data->onReceive != LUA_NOREF) {
+    lua_rawgeti(L, LUA_REGISTRYINDEX, data->onReceive); // load the callback function
+    lua_rawgeti(L, LUA_REGISTRYINDEX, data->self_ref);  // pass itself, #1 callback argument
+    lua_pushstring(L, message); // #2 callback argument
+    lua_pushnumber(L, opCode); // #3 callback argument
+    lua_call(L, 3, 0);
+  }
+}
+
+static void websocketclient_onCloseCallback(ws_info *ws, int errorCode) {
+  NODE_DBG("websocketclient_onCloseCallback\n");
+
+  lua_State *L = lua_getstate();
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    luaL_error(L, "Client websocket is nil.\n");
+    return;
+  }
+  ws_data *data = (ws_data *) ws->reservedData;  
+
+  if (data->onClose != LUA_NOREF) {
+    lua_rawgeti(L, LUA_REGISTRYINDEX, data->onClose); // load the callback function
+    lua_rawgeti(L, LUA_REGISTRYINDEX, data->self_ref);  // pass itself, #1 callback argument
+    lua_pushnumber(L, errorCode); // pass the error code, #2 callback argument
+    lua_call(L, 2, 0);
+  }
+
+  // free self-reference to allow gc (no futher callback will be called until next ws:connect())
+  lua_gc(L, LUA_GCSTOP, 0); // required to avoid freeing ws_data
+  luaL_unref(L, LUA_REGISTRYINDEX, data->self_ref);
+  data->self_ref = LUA_NOREF;
+  lua_gc(L, LUA_GCRESTART, 0);
+}
+
+static int websocket_createClient(lua_State *L) {
+  NODE_DBG("websocket_createClient\n");
+
+  // create user data
+  ws_data *data = (ws_data *) luaM_malloc(L, sizeof(ws_data));
+  data->onConnection = LUA_NOREF;
+  data->onReceive = LUA_NOREF;
+  data->onClose = LUA_NOREF;
+  data->self_ref = LUA_NOREF; // only set when ws:connect is called
+
+  ws_info *ws = (ws_info *) lua_newuserdata(L, sizeof(ws_info));
+  ws->connectionState = 0;
+  ws->onConnection = &websocketclient_onConnectionCallback;
+  ws->onReceive = &websocketclient_onReceiveCallback;
+  ws->onFailure = &websocketclient_onCloseCallback;
+  ws->reservedData = data;
+
+  // set its metatable
+  luaL_getmetatable(L, METATABLE_WSCLIENT);
+  lua_setmetatable(L, -2);
+
+  return 1;
+}
+
+static int websocketclient_onConnection(lua_State *L) {
+  NODE_DBG("websocketclient_onConnection\n");
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+  luaL_argcheck(L, ws, 1, "Client websocket expected");
```

Comment:
All checks for `ws` after `luaL_checkudata()` are redundant since this function will effectively not return to your context in case of an error. It'll long jump to the calling Lua context instead. Ref. https://www.lua.org/manual/5.1/manual.html#lua_error.
Also applies to similar instances below.